app/models/membership_application.rb

```diff
@@ -136,11 +136,13 @@ def is_member?
   def accept_membership
     begin
 
+      membership_number = self.membership_number.blank? ? self.class.get_next_membership_number : self.membership_number
+
       company = Company.find_or_create_by!(company_number: company_number) do |co|
         co.email = contact_email
       end
 
-      update(company: company)
+      update(company: company, membership_number: membership_number)
```

Comment:
Perhaps line 39 should be immediately before this line.

Also, should we wrap those two lines in a transaction?  We have a small window of vulnerability here - but I'm not sure its worth worrying about as it seems highly unlikely to lead to a collision.

Reply:
In my view, there would only be a change of collision if the `get_next_membership_number` would depend on the record being updated (e.g. if we would get the value by querying the database for max number + 1) . But that's precisely what the sequence is meant to solve: it will never give out the same number twice. Am I overlooking something?

Reply:
You're probably right about this and it's just my ignorance that's getting in the way.

This statement in the `select` for sequence number ....
```
setval('membership_number_seq', (SELECT COALESCE(MAX(CAST(membership_number AS INT))+1, ...
```
... looks like it will return the _same_ sequence number for two different sessions if executed against the same `membership_number`.   I guess (?) it would actually return `membership_number+1` and `membership_number+2` if called twice again the same, unchanged table.  Is that how it works?

Reply:
The `#setval` method for the sequence is just an initialisation method. It sets the starting value for the sequence. It is not a definition for the way the sequence numbers are generated.

So after the sequence is set up by the migration it is just an independent number generator, that gives out values incremented by 1.

Reply:
Ah, I see .... thanks for the explanation.

And, of course you're right - we're safe from colliding on the same number.